### PR TITLE
dev-util/rpi-imager: Add new package

### PR DESCRIPTION
**Pull request by martin-kokos**
rpi-imager is a utility for creating bootable images for popular RaspberryPi computers - https://www.raspberrypi.com/products/

### CODE REVIEW

**Comment by pastalian** [comment]
How did you create this list of dependencies? I feel that many of them are unnecessary and need  to be cleanup.

**Comment by martin-kokos** [comment]
> How did you create this list of dependencies? I feel that many of them are unnecessary and need to be cleanup.

I used the command `ldd /usr/bin/rpi-imager|cut -d' ' -f 3|xargs equery b | cut -d' ' -f 1 | sort -u | cut -d'-' -f 1-2` which I thought would capture all rdeps.

**Comment by pastalian** [comment]
You should try `scanelf -n /usr/bin/rpi-imager` to find out which are real rdeps and which are transitive dependencies.

Ideally, check CMakeLists.txt as well. There should be room for improvement, such as adding a USE flag to select qt5/6.
https://github.com/raspberrypi/rpi-imager/blob/qml/src/CMakeLists.txt#L97-L104

File: dev-util/rpi-imager/rpi-imager-1.8.5.ebuild [hunk 1]
```diff
@@ -0,0 +1,66 @@
+# Copyright 2024 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=8
+
+inherit cmake
```
Comment (pastalian) [inline dev-util/rpi-imager/rpi-imager-1.8.5.ebuild:6]:
```suggestion
inherit cmake xdg
```
Inherit xdg eclass to update icon/desktop file caches.

File: dev-util/rpi-imager/rpi-imager-1.8.5.ebuild [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+https://www.raspberrypi.com/documentation/computers/getting-started.html#raspberry-pi-imager
+https://github.com/raspberrypi/rpi-imager
+"
+SRC_URI="https://github.com/raspberrypi/rpi-imager/archive/refs/tags/v${PV}.tar.gz -> ${P}.tar.gz"
+S="${WORKDIR}/${P}/src"
+
+LICENSE="Apache-2.0"
+SLOT="0"
+
+KEYWORDS="~amd64"
```
Comment (pastalian) [inline dev-util/rpi-imager/rpi-imager-1.8.5.ebuild:20]: Add USE flags to handle conditional dependencies. Without these flags and src_configure, this ebuild is non-deterministic.
```
IUSE="gnutls qt6"
```

File: dev-util/rpi-imager/rpi-imager-1.8.5.ebuild [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+RDEPEND="
```
Comment (pastalian) [inline dev-util/rpi-imager/rpi-imager-1.8.5.ebuild:22]: I would start with something like this. It's not perfect and likely missing something. But as I said, putting everything that comes out of ldd into RDEPND is a bit rough.
```
RDEPEND="
	app-arch/libarchive:=
	app-arch/xz-utils
	net-misc/curl
	sys-apps/util-linux
	sys-libs/zlib
	gnutls? ( net-libs/gnutls:= )
	!gnutls? ( dev-libs/openssl:= )
	qt6? (
		dev-qt/qtbase:6[concurrent,dbus,gui,network,widgets]
		dev-qt/qtdeclarative:6[svg]
	)
	!qt6? (
		dev-qt/qtconcurrent:5
		dev-qt/qtcore:5
		dev-qt/qtdbus:5
		dev-qt/qtdeclarative:5
		dev-qt/qtgui:5
		dev-qt/qtnetwork:5
		dev-qt/qtquickcontrols2:5
		dev-qt/qtsvg:5
		dev-qt/qtwidgets:5
	)
"
DEPEND="${RDEPEND}"
BDEPEND="
	qt6? ( dev-qt/qttools:6[linguist] )
	!qt6? ( dev-qt/linguist-tools:5 )
"
```

File: dev-util/rpi-imager/rpi-imager-1.8.5.ebuild [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+	sys-apps/acl
+	sys-apps/dbus
+	sys-devel/gcc
+	sys-libs/glibc
+	sys-libs/zlib
+	x11-libs/libX11
+	x11-libs/libXau
+	x11-libs/libxcb
+	x11-libs/libXdmcp
+"
```
Comment (pastalian) [inline dev-util/rpi-imager/rpi-imager-1.8.5.ebuild:66]: [gnutls / openssl](https://github.com/raspberrypi/rpi-imager/blob/4b6ee9bb4b1f29587db6145d82f4784d8863d036/src/CMakeLists.txt#L42)
[qt5 / qt6](https://github.com/raspberrypi/rpi-imager/blob/4b6ee9bb4b1f29587db6145d82f4784d8863d036/src/CMakeLists.txt#L97)
`cmake_use_find_package` can handle these situations.
```
src_configure() {
	local mycmakeargs=(
		$(cmake_use_find_package gnutls GnuTLS)
		$(cmake_use_find_package !qt6 Qt5)
	)
	cmake_src_configure
}
```

File: dev-util/rpi-imager/rpi-imager-1.8.5.ebuild (hunk 3 above)
Comment (martin-kokos) [reply dev-util/rpi-imager/rpi-imager-1.8.5.ebuild:22]: Thanks for the help. I am looking into how you did these so far and will look into CMakeLists more.

Comment (pastalian) [reply dev-util/rpi-imager/rpi-imager-1.8.5.ebuild:22]: A quick glance at CMakeLists will give you a general idea of the dependencies. 
For example,
https://github.com/raspberrypi/rpi-imager/blob/4b6ee9bb4b1f29587db6145d82f4784d8863d036/src/CMakeLists.txt#L42-L50
need gnutls _or_ openssl

https://github.com/raspberrypi/rpi-imager/blob/4b6ee9bb4b1f29587db6145d82f4784d8863d036/src/CMakeLists.txt#L97-L104
need qt5 _or_ qt6

https://github.com/raspberrypi/rpi-imager/blob/4b6ee9bb4b1f29587db6145d82f4784d8863d036/src/CMakeLists.txt#L331
need curl

https://github.com/raspberrypi/rpi-imager/blob/4b6ee9bb4b1f29587db6145d82f4784d8863d036/src/CMakeLists.txt#L335
need lsblk (util-linux)

File: dev-util/rpi-imager/rpi-imager-1.8.5.ebuild [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+		dev-qt/qtwidgets:5
+	)
+"
+
+DEPEND="${RDEPEND}"
+BDEPEND="
+	gnutls? ( net-libs/gnutls:= )
+	!gnutls? ( dev-libs/openssl:= )
+	net-misc/curl
+	app-arch/libarchive:=
```
Comment (pastalian) [inline dev-util/rpi-imager/rpi-imager-1.8.5.ebuild:82]: BDEPEND is for build-time dependencies (what you need to execute during the build). I believe these are used as libraries. If so, they should be RDEPEND.

File: dev-util/rpi-imager/rpi-imager-1.8.5.ebuild [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+RDEPEND="
+	sys-apps/util-linux
+	zlib? ( sys-libs/zlib )
+	lzma? ( app-arch/lzma )
+	zstd? ( app-arch/zstd )
+	brotli? ( app-arch/brotli )
+	bzip2? ( app-arch/bzip2 )
+	xz? ( app-arch/xz-utils )
+	dbus? ( sys-apps/dbus )
```
Comment (pastalian) [inline dev-util/rpi-imager/rpi-imager-1.8.5.ebuild:59]: You don't need to fix it in this PR, but for future improvement: these USE flags are not very useful as they are automagic dependencies.
https://wiki.gentoo.org/wiki/Project:Quality_Assurance/Automagic_dependencies

File: dev-util/rpi-imager/rpi-imager-1.8.5.ebuild [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+https://github.com/raspberrypi/rpi-imager
+"
+SRC_URI="https://github.com/raspberrypi/rpi-imager/archive/refs/tags/v${PV}.tar.gz -> ${P}.tar.gz"
+S="${WORKDIR}/${P}/src"
+
+LICENSE="Apache-2.0"
+SLOT="0"
+
+KEYWORDS="~amd64"
+IUSE="gnutls qt5 qt6 zlib lzma zstd brotli bzip2 xz dbus gnutls openssl -telemetry"
```
Comment (pastalian) [inline dev-util/rpi-imager/rpi-imager-1.8.5.ebuild:20]:
```suggestion
IUSE="gnutls qt5 qt6 zlib lzma zstd brotli bzip2 xz dbus gnutls openssl telemetry"
```
Do not add -. It doesn't do anything.

**Review by pastalian — COMMENTED** [review]
Thanks for the adjustment! I have made a few final comments.

File: dev-util/rpi-imager/rpi-imager-1.8.5.ebuild [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+	x11-libs/libX11
+	x11-libs/libXau
+	x11-libs/libxcb
+	x11-libs/libXdmcp
+"
+
+RDEPEND="
+	sys-apps/util-linux
+	zlib? ( sys-libs/zlib )
+	lzma? ( app-arch/lzma )
```
Comment (pastalian) [inline dev-util/rpi-imager/rpi-imager-1.8.5.ebuild:54]: It's confusing, but liblzma is not app-arch/lzma. it's app-arch/xz-utils.

File: dev-util/rpi-imager/rpi-imager-1.8.5.ebuild (hunk 6 above)
Comment (pastalian) [inline dev-util/rpi-imager/rpi-imager-1.8.5.ebuild:59]: I don't think these USE flag do anything. If not, ignore me.

File: dev-util/rpi-imager/rpi-imager-1.8.5.ebuild (hunk 7 above)
Comment (martin-kokos) [reply dev-util/rpi-imager/rpi-imager-1.8.5.ebuild:20]: I was reading this
>  Add + or - before the name of the use flag in IUSE to turn it on or off by default. 
https://devmanual.gentoo.org/general-concepts/use-flags/#iuse-defaults
but I guess its - by default.

Comment (pastalian) [reply dev-util/rpi-imager/rpi-imager-1.8.5.ebuild:20]: Yes. Noticed that it says "Important: Adding - before a flag in IUSE is pretty much useless."

Comment (martin-kokos) [reply dev-util/rpi-imager/rpi-imager-1.8.5.ebuild:20]: Yes. I guess I saw that but the point didn't come across very clearly.

File: dev-util/rpi-imager/rpi-imager-1.8.5.ebuild (hunk 6 above)
Comment (martin-kokos) [reply dev-util/rpi-imager/rpi-imager-1.8.5.ebuild:59]: > The imaging utility will normally be run as regular user, and will call udisks2 over DBus to perform privileged operations like opening the disk device for writing.
If udisks2 is not functional on your Linux distribution, you can alternatively start it as "root" with sudo and similar tools.

https://github.com/raspberrypi/rpi-imager/blob/qml/README.md?plain=1#L62

Comment (pastalian) [reply dev-util/rpi-imager/rpi-imager-1.8.5.ebuild:59]: But it uses dbus through QT, which you already added (dev-qt/qtbase[dbus] for qt6, dev-qt/qtdbus for qt5).

**Comment by pastalian** [comment]
@martin-kokos Hi. Are you still working on this? If not, I can merge this with some followups.

**Comment by martin-kokos** [comment]
> @martin-kokos Hi. Are you still working on this? If not, I can merge this with some followups.

I was going to verify the dbus functionality through QT with d-spy, but didn't get around to it yet, but otherwise I didn't find any other issues.

**Comment by pastalian** [comment]
Take your time :) Let me known when you think it's ready.

**Comment by pastalian** [comment]
I went ahead and merged it at f88eabe with some followups 3958f7b (these are not in the master branch yet).
Thank you for the contribution!

btw, I recommend that you request commit access so that you can commit directly to the dev branch.
https://wiki.gentoo.org/wiki/Project:GURU/Information_for_Contributors

**Pull request closed without merging**